### PR TITLE
Added support for writing non-triangular meshes in .OBJ files -- GH #1155

### PR DESCRIPTION
**Pull request by glyg**
Here we go.

Not sure the tests are correctly or sensibly written. Also not sure if one can (and should) test if a warning is raised.

Waiting for comments!

### CODE REVIEW

File: vispy/io/wavefront.py [hunk 1]
```diff
@@ -303,10 +305,13 @@ def writeMesh(self, vertices, faces, normals, values, name=''):
...
-        Nfaces = faces.size // 3
-        faces = faces.reshape((Nfaces, 3))
-
+        try:
+            # Reshape faces
+            Nfaces = faces.size // 3
+            faces = faces.reshape((Nfaces, 3))
+        except ValueError:
+            logger.warning('''Faces doesn't appear to be triangular, '''
+                           '''be advised the file cannot be read back''')
```
Comment (almarklein) [inline vispy/io/wavefront.py]: Can you please add "... be read back *in Vispy*" here, to make it explicit that this is a vispy thing?

**Comment by almarklein** [comment]
Other than that, this LGTM

**Comment by glyg** [comment]
Fixed the warning and squashed the commits

Thx for the review

File: vispy/io/wavefront.py [hunk 2]
```diff
@@ -303,10 +305,14 @@ def writeMesh(self, vertices, faces, normals, values, name=''):
...
             faces = np.arange(len(vertices))
 
-        # Reshape faces
-        Nfaces = faces.size // 3
-        faces = faces.reshape((Nfaces, 3))
-
+        try:
+            # Reshape faces
+            Nfaces = faces.size // 3
+            faces = faces.reshape((Nfaces, 3))
```
Comment (Eric89GXL) [inline vispy/io/wavefront.py]: hmm... but couldn't you have a non-triangular file that just happened to be divisibly by 3, thus making this "check" not correct?

**Comment by glyg** [comment]
Damn you're right!
Now I'm confused.

My understanding is that the reshape has to happen primarily if the `faces` argument was not provided, in which case it's created just above the reshape, and needs reshaping... Now if the user provides the `face` argument, one can assume it will be correctly shaped? 

Said otherwise, if `faces` is already there, why reshape it without telling the user?

Another point to consider: in the provided test, `faces` is a 1D array of lists, so it's even worse, because you'll end up with strange triples of lists after the reshape, which will raise a `TypeError` further down the road, when increasing the index by 1...

**Comment by Eric89GXL** [comment]
I am a bit far removed from the code at this point to be able to tell you definitively. Can you tell from the tests what is supposed to happen?

Instead of implicitly trying to reshape, could we add or modify a `kwarg` that would default to doing a reshape (to match current behavior) but could be turned off (for your use case)?

**Comment by glyg** [comment]
I rewrote the test with three polygonal faces, and it does rise a `TypeError`

```python
/home/guillaume/anaconda3/envs/py35/lib/python3.5/site-packages/vispy-0.5.0.dev0-py3.5.egg/vispy/io/wavefront.py in writeFace(self, val, what)
    278         """
    279         # OBJ counts from 1
--> 280         val = [v + 1 for v in val]
    281         # Make string
    282         if self._hasValues and self._hasNormals:

/home/guillaume/anaconda3/envs/py35/lib/python3.5/site-packages/vispy-0.5.0.dev0-py3.5.egg/vispy/io/wavefront.py in <listcomp>(.0)
    278         """
    279         # OBJ counts from 1
--> 280         val = [v + 1 for v in val]
    281         # Make string
    282         if self._hasValues and self._hasNormals:

TypeError: can only concatenate list (not "int") to list
```

If a `kwarg` is acceptable, it is definitely the easier way to go.
I'll modify the code to see how it rolls.

File: vispy/io/wavefront.py (hunk 1 above)
Comment (QuLogic) [inline vispy/io/wavefront.py]: Also, why waste the space for triple quotes if you aren't going to use its main benefit (unescaped newlines).

Comment (glyg) [inline vispy/io/wavefront.py]: fair point! I'll fix that too

**Comment by Eric89GXL** [comment]
Thanks @glyg

**Comment by glyg** [comment]
thanks to you all! Hope I can be of more use in the future!

**Pull request merged**